File: Go/test-integration/tests/recursion.go
func main() {
  var n = 5;

  println(fact(n - 1));  // should print "24"
  println(fact(n * 2));  // should print "3628800"
  println(fact(n));      // should print "120"
  println(fib(n));       // should print "8"
  println(foo(1000));    // should print "0"
  println(bar(1000));    // should print "0"
  println(fib(fact(3))); // should print "13"
}


/* Factorial */

func fact(n int) int {
  if n == 0 || n == 1 { return 1; };
  return n * fact(n - 1);
}


/* Fibonacci numbers */

func fib(n int) int {
  if n == 0 || n == 1 {
    return 1;
  };
  return fib(n - 1) + fib(n - 2);
}


/* Mutual recursion */

func foo(n int) int {
  if n == 0 { return 0; };
  return bar(n - 1);
}

func bar(n int) int {
  if n == 0 { return 0; };
  return foo(n - 1);
}

File: FSharpUoM/test-integration/tests/recursion.fs
let rec fact n = if n <= 1 then 1 else n * fact (n - 1);
fact 5;; // Right: "120"

let n = 15;
let rec fact n = if n <= 1 then 1 else n * fact (n - 1);
fact (n - 2);; // Right: "6227020800"

let rec fib n = if n <= 1 then 1 else fib (n - 1) + fib (n - 2);
fib 5;; // Right: "8"

let n = 30;
let rec fib n = if n <= 1 then 1 else fib (n - 1) + fib (n - 2);
fib (n - 2);; // Right: "514229"

let rec f x = if x = 0 then 0 else begin
    let g x = if x = 0 then 0 else f (x - 1) in
        g (x - 1)
end;
f 5;; // Right: "0"